src/Data/Foreign/Generic.purs

```diff
@@ -101,21 +110,26 @@ readGeneric { sumEncoding
         y <- go (_2 unit) b
         pure $ SProd "Data.Tuple.Tuple" [\_ -> x, \_ -> y]
       _ -> Left (TypeMismatch "array of length 2" "array")
+  go (SigProd _ alts) f | untagEnums && all (\a -> length a.sigValues == 0) alts = do
+    tag <- readString f
+    case find (\alt -> (constructorTagModifier alt.sigConstructor) == tag) alts of
+      Nothing -> Left (TypeMismatch ("one of " <> show (map (constructorTagModifier <<< _.sigConstructor ) alts)) tag)
```

Comment:
While we're breaking the API, it might be a good time to change `TypeMismatch` so that it takes an array of alternatives as an argument instead of a `String`.